boogie_ast/src/boogie_program/writer.rs

```diff
@@ -0,0 +1,320 @@
+                writer.indent()?;
+                writeln!(writer.writer, "{label}:")?;
+            }
+            Stmt::Return => {
+                writer.indent()?;
+                writeln!(writer.writer, "return;")?;
+            }
+            Stmt::While { condition, body } => {
+                writer.indent()?;
+                write!(writer.writer, "while (")?;
+                writer.write(condition)?;
+                writeln!(writer.writer, ") {{")?;
+                writer.increase_indent();
+                writer.write(body.as_ref())?;
+                writer.decrease_indent();
+                writeln!(writer.writer, "}}")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Writable for Contract {
+    fn write_to<T: Write>(&self, writer: &mut Writer<T>) -> std::io::Result<()> {
+        for r in &self.requires {
+            writer.indent()?;
+            write!(writer.writer, "requires ")?;
+            writer.write(r)?;
+            writeln!(writer.writer, ";")?;
+        }
+        for e in &self.ensures {
+            writer.indent()?;
+            write!(writer.writer, "ensures ")?;
+            writer.write(e)?;
+            writeln!(writer.writer, ";")?;
+        }
+        for m in &self.modifies {
+            writer.indent()?;
+            write!(writer.writer, "modifies ")?;
+            writer.write(m)?;
+            writeln!(writer.writer, ";")?;
+        }
+        Ok(())
+    }
+}
+
+impl Writable for Type {
+    fn write_to<T: Write>(&self, writer: &mut Writer<T>) -> std::io::Result<()> {
+        match self {
+            Type::Bool => write!(writer.writer, "bool")?,
+            Type::Bv(size) => write!(writer.writer, "bv{size}")?,
+            Type::Int => write!(writer.writer, "int")?,
+            Type::Map { key, value } => {
+                write!(writer.writer, "[")?;
+                writer.write(key.as_ref())?;
+                write!(writer.writer, "]")?;
+                writer.write(value.as_ref())?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Writable for Literal {
+    fn write_to<T: Write>(&self, writer: &mut Writer<T>) -> std::io::Result<()> {
+        match self {
+            Literal::Bool(value) => {
+                write!(writer.writer, "{}", value)?;
+            }
+            Literal::Bv { width, value } => {
+                write!(writer.writer, "{value}bv{width}")?;
+            }
+            Literal::Int(value) => {
+                write!(writer.writer, "{}", value)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Writable for UnaryOp {
+    fn write_to<T: Write>(&self, writer: &mut Writer<T>) -> std::io::Result<()> {
+        match self {
+            UnaryOp::Not => write!(writer.writer, "!")?,
+            UnaryOp::Neg => write!(writer.writer, "-")?,
+        }
+        Ok(())
+    }
+}
+
+impl Writable for BinaryOp {
+    fn write_to<T: Write>(&self, writer: &mut Writer<T>) -> std::io::Result<()> {
```

Comment:
Wouldn't it make sense to implement Display instead?

Reply:
Are you referring to implementing it for `BinaryOp` specifically, or for everything? For `BinaryOp`, I think it makes sense to implement it since it's a few characters long, and it could help with debugging.

Reply:
I think everything

Reply:
The reason I avoided using `Display` is that I wanted to maximize the buffering of the output and minimize the creation of intermediate strings that represent components of the Boogie program. The idea is that since everything gets written to a file, we can pass the file buffer to every `write_to` method to allow them to write to it directly as opposed to having to create intermediate strings (which can be large), and then writing them to the file. For instance, if we were to implement `Display` for the `Contract` type, then `Procedure`'s `write_to` method will do:
```rust
write!(writer, "{contract}");
```
which will convert the whole contract into a string, and then write it to the buffer. Whereas, the current approach:
```rust
contract.write_to(writer);
```
allows the components of the contract to be written directly to the buffer without creating an intermediate string that represents the whole contract.

Does this make sense?